rskj-core/src/main/java/co/rsk/peg/pegininstructions/PeginInstructionsProvider.java

```diff
@@ -0,0 +1,85 @@
+package co.rsk.peg.pegininstructions;
+
+import co.rsk.bitcoinj.core.BtcTransaction;
+import co.rsk.bitcoinj.script.ScriptChunk;
+import co.rsk.bitcoinj.script.ScriptOpCodes;
+import co.rsk.peg.NoOpReturnException;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PeginInstructionsProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(PeginInstructionsProvider.class);
+
+    public Optional<PeginInstructions> buildPeginInstructions(BtcTransaction btcTx) throws
+        PeginInstructionsException {
+
+        PeginInstructions peginInstructions;
+        byte[] opReturnOutputData;
+
+        try {
+            opReturnOutputData = extractOpReturnData(btcTx);
+        } catch (NoOpReturnException e) {
+            return Optional.empty();
+        }
+
+        if (opReturnOutputData == null || opReturnOutputData.length < 22) {
```

Comment:
These validations don't seem appropriate for this class

Reply:
It could be on method extractOpReturnData, but it is still on this class. Where should it be so?

Reply:
just left another comment about this. You already check the length in the base class of PeginInstructions. You shouldn't duplicate that logic.
Closing this comment as it's repeated :)